warehouse/utils/security_policy.py

```diff
@@ -0,0 +1,118 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import enum
+
+from pyramid.authorization import Authenticated
+from pyramid.interfaces import ISecurityPolicy
+from pyramid.security import Denied
+from zope.interface import implementer
+
+from warehouse.accounts.interfaces import IUserService
+from warehouse.accounts.models import User
+
+
+class AuthenticationMethod(enum.Enum):
+    BASIC_AUTH = "basic-auth"
+    SESSION = "session"
+    MACAROON = "macaroon"
+
+
+def _principals_for_authenticated_user(user, request):
+    """Apply the necessary principals to the authenticated user"""
+    login_service = request.find_service(IUserService, context=None)
+
+    if request.session.password_outdated(login_service.get_password_timestamp(user.id)):
+        request.session.invalidate()
+        request.session.flash("Session invalidated by password change", queue="error")
+        return None
```

Comment:
This was part of the original code so I haven't moved it, but it's oddly placed here -- it's really part of AuthN, not the AuthZ work we're doing here. Any objections to me moving it into the `identity`/AuthN phase instead?

Reply:
Done with [29b40f9](https://github.com/pypa/warehouse/pull/11218/commits/29b40f969b7f5191b89189f82b061854295226cf)